src/c1_koiter_steigmann/koiter_steigmann_curvable_bell_elements.h

```diff
@@ -9,580 +9,1183 @@
-                  // Fill in the affine part of the rotation of the second
-                  // derivatives [zdec] if( beta>= alpha) ?
-                  b22(row_b22, col) += jac_inv(mu, alpha) * jac_inv(nu, beta);
-                }
-              }
+              left_data_used[left_duid1_index] = true;
+              continue;
             }
           }
-        }
-      }
-
-      // Fill in the submatrices to the full (6x6) matrix
-      Rotation_matrix_at_node[j_node](0, 0) = 1.0;
-      // Fill in b1 --- the affine contribution to rotation of the
-      // first derivatives
-      for (unsigned i = 0; i < 2; ++i)
-      {
-        for (unsigned j = 0; j < 2; ++j)
-        {
-          Rotation_matrix_at_node[j_node](1 + i, 1 + j) = b1(i, j);
-        }
-      }
-      // Fill in b21 --- the non-affine (second derivative dependent)
-      // rotation of the first derivatives
-      for (unsigned i = 0; i < 2; ++i)
-      {
-        for (unsigned j = 0; j < 3; ++j)
-        {
-          Rotation_matrix_at_node[j_node](1 + i, 3 + j) = b12(i, j);
-        }
-      }
-      // Fill in b22 --- the rotation of the second derivatives
-      for (unsigned i = 0; i < 3; ++i)
-      {
-        for (unsigned j = 0; j < 3; ++j)
-        {
-          Rotation_matrix_at_node[j_node](3 + i, 3 + j) = b22(i, j);
-        }
-      }
-    } // end fill_in_rotation_matrix_at_node_with_basis
-
-    /// Pointer to the `parent' finite element which this is a helper force
-    FiniteElement* Parent_element_pt;
-
-    /// The number of nodes (that we store rotation data for) in the fvk element
-    /// that uses this helper
-    unsigned Nnode;
-
-    /// Vector containing boundary parametrised location for each node
-    Vector<double> Boundary_coordinate_of_node;
+          if (!left_duid2_pinned && !left_data_used[left_duid2_index])
+          {
+            // [zdec] debug
+            // std::cout << "eqn " << condition_index << " depends on dof L"
+            //        << left_duid2_index << std::endl;
+            double coeff = jac_of_transform(1, alpha);
+            if (fabs(coeff) > Orthogonality_tolerance)
+            {
+              left_data_used[left_duid2_index] = true;
+              continue;
+            }
+          }
+          // ---------------------------------------------------------------------
+          // If we made it here, it is because all dofs in the constraint are
+          // pinned so we need to check the constraint is satisfied manually and
+          // then remove it by pinning the corresponding lagrange multiplier
 
-    /// Vector containing boundary parametrisation at each node
-    Vector<CurvilineGeomObject*> Nodal_boundary_parametrisation_pt;
-
-    /// Vector containing <rotation matrix at each node>
-    Vector<DenseMatrix<double>> Rotation_matrix_at_node;
-  };
-
-  //---end of rotation helper class-------------------------------------------
+          // hierher Aidan should this stay alive?
```

Comment:
Answered earlier

Reply:
Aidan: please tidy up brutally; I'll keep this comment in there to remind you but will mark this as resolved